esp-hal/src/aes/mod.rs

```diff
@@ -449,23 +450,29 @@ pub mod dma {
             iv[12..16].copy_from_slice(&self.aes.regs().iv_mem(3).read().bits().to_le_bytes());
         }
 
-        fn start_dma_transfer(
-            &mut self,
+        fn start_dma_transfer<RXBUF, TXBUF>(
+            mut self,
             number_of_blocks: usize,
-            output: &mut impl DmaRxBuffer,
-            input: &mut impl DmaTxBuffer,
+            mut output: RXBUF,
+            mut input: TXBUF,
             mode: Mode,
             cipher_mode: CipherMode,
             key: &Key,
-        ) -> Result<(), DmaError> {
+        ) -> Result<AesTransfer<'d, RXBUF, TXBUF>, (DmaError, Self, RXBUF, TXBUF)>
+        where
+            TXBUF: DmaTxBuffer,
+            RXBUF: DmaRxBuffer,
+        {
             let peri = self.dma_peripheral();
 
-            unsafe { self.channel.tx.prepare_transfer(peri, input) }?;
-            unsafe { self.channel.rx.prepare_transfer(peri, output) }?;
-
-            // Start them together, to avoid the latter prepare discarding data from FIFOs.
-            self.channel.tx.start_transfer()?;
-            self.channel.rx.start_transfer()?;
+            if let Err(error) = unsafe { self.channel.tx.prepare_transfer(peri, &mut input) }
+                .and_then(|_| unsafe { self.channel.rx.prepare_transfer(peri, &mut output) })
+                // Start them together, to avoid the latter prepare discarding data from FIFOs.
+                .and_then(|_| self.channel.tx.start_transfer())
+                .and_then(|_| self.channel.rx.start_transfer())
```

Comment:
It doesn't need fixing in this PR but regarding DMA APIs in general, we need to stop the TX channel if the RX channel fails to start. Just so we don't have a half started DMA just lingering around.